0.13.0 is no longer compatible with OSX 10.7

Reported by rancor on IRC #bitcoin , reproduced on my 10.7VM.
Works fine on 10.8.

`InitSanityCheck()` fails, ... there is no `ECC_InitSanityCheck()` log output which means, `glibc_sanity_test()` or `glibcxx_sanity_test()` are failing.

Either we analyze and fix it for 0.13.1 or we drop support for OSX 10.7.

Ok, bad news. I ended up in the bowels of osx, but I have somewhat of a handle on this now.

- In past releases (pre-0.13) we always used libstdc++, which works everywhere
- libstdc++ contains only minimal c++11 support. No atomics, for example.
- libc++ was introduced in 10.7 and made the default in 10.9. It's the only option for full c++11 support.

From what I can tell, 10.7's runtime is buggy in several ways. Or it may be that there are some undefined conditions that it reacts to differently in 10.8+.

Here's an example of one that crashes in 10.7 but works elsewhere. This is a simplified version of an actual failing test of ours:
```c++
#include <assert.h>
#include <sstream>

int main()
{
    std::istringstream text("1e10000");
    double result;
    text >> result;
    assert(text.fail());
    return 0;
}
```
It appears as though there are several issues relating to text conversion and localization. I didn't manage to track down the typeinfo bug that @jonasschnelli pointed out above, but I assume it comes from a small abi difference in 10.8+.

I tried all kinds of build combinations, but we're too entrenched in c++11 to make libstdc++ work.

I've also been able to confirm the above by grabbing my 10.11 libc++ dylibs and forcing 10.7 to use them. These are runs of the test_bitcoin from gitian, on a 10.7 vm:
```bash
corys-MacBook-Pro:bin cory$ ./test_bitcoin 
Running 212 test cases...
test/uint256_tests.cpp:197: error: in "uint256_tests/methods": check ss.str() == std::string(ZeroArray,ZeroArray+32) has failed
test/uint256_tests.cpp:199: error: in "uint256_tests/methods": check ZeroL == TmpL has failed
test/uint256_tests.cpp:202: error: in "uint256_tests/methods": check ss.str() == std::string(MaxArray,MaxArray+32) has failed
test/uint256_tests.cpp:204: error: in "uint256_tests/methods": check MaxL == TmpL has failed
test/uint256_tests.cpp:237: error: in "uint256_tests/methods": check ss.str() == std::string(R1Array,R1Array+20) has failed
test/uint256_tests.cpp:242: error: in "uint256_tests/methods": check ss.str() == std::string(ZeroArray,ZeroArray+20) has failed
test/uint256_tests.cpp:244: error: in "uint256_tests/methods": check ZeroS == TmpS has failed
test/uint256_tests.cpp:247: error: in "uint256_tests/methods": check ss.str() == std::string(MaxArray,MaxArray+20) has failed
test/uint256_tests.cpp:249: error: in "uint256_tests/methods": check MaxS == TmpS has failed
test/util_tests.cpp:466: error: in "util_tests/test_ParseDouble": check !ParseDouble("-1e10000", __null) has failed
test/util_tests.cpp:467: error: in "util_tests/test_ParseDouble": check !ParseDouble("1e10000", __null) has failed

*** 11 failures are detected in the test module "Bitcoin Test Suite"
```
And the same binary, but with a hijacked libc++/libc++abi:
```bash
corys-MacBook-Pro:bin cory$ DYLD_LIBRARY_PATH=`pwd` ./test_bitcoin 
Running 212 test cases...

*** No errors detected
```
bitcoin-qt works fine that way too.

So... I don't see any good solution here. 0.13.0 must be broken as well, so it wouldn't be a regression to release without 10.7 support. The only viable option would be to package a known-good libc++/libc++abi in the dmg, which is commonly done, but has a number of implications for us this late in the release cycle (namely signing related).

--- a/doc/build-osx.md
+++ b/doc/build-osx.md
@@ -90,6 +90,6 @@ Uncheck everything except Qt Creator during the installation process.
 Notes
 -----
 
-* Tested on OS X 10.7 through 10.11 on 64-bit Intel processors only.
+* Tested on OS X 10.8 through 10.12 on 64-bit Intel processors only.
 
 * Building with downloaded Qt binaries is not officially supported. See the notes in [#7714](https://github.com/bitcoin/bitcoin/issues/7714)

--- a/doc/release-notes.md
+++ b/doc/release-notes.md
@@ -36,6 +36,10 @@ No attempt is made to prevent installing or running the software on Windows XP,
 you can still do so at your own risk, but do not expect it to work: do not
 report issues about Windows XP to the issue tracker.
 
+From 0.13.1 onwards OS X 10.7 is no longer supported. 0.13.0 was intended to work on 10.7+, 
+but severe issues with the libc++ version on 10.7.x keep it from running reliably. 
+0.13.1 now requires 10.8+, and will communicate that to 10.7 users, rather than crashing unexpectedly.
+
 Notable changes
 ===============
 

